### PR TITLE
Added simple transfer with memo

### PR DESCRIPTION
**Pull request by shjortConcordium**
## Purpose

Add support for simple transfers with memo.

## Changes

Added Simple transfer with memo type and handler.
Improved typing of getAccountTransactionHandler function.
Renamed SimpleTransfer to SimpleTransferPayload.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -49,11 +46,42 @@ const simpleTransferAccountTransaction: AccountTransaction = {
...
+const simpleTransferWithMemoAccountTransaction: AccountTransaction = {
+    header: header,
+    payload: simpleTransferWithMemo,
+    type: AccountTransactionType.SimpleTransferWithMemo,
+};
+```
+
+## Send Account Transaction
+The following example demonstrates how to send any account transaction.
+See previous sections how to create one.
```
Comment (orhoj) [inline README.md:73]: See the previous sections for how to create an account transaction.

File: src/types.ts [hunk 2]
```diff
@@ -453,15 +457,21 @@ export interface AccountTransactionHeader {
...
-    /** the recipient of the transfer*/
+    /** the recipient of the transfer */
     toAddress: AccountAddress;
 }
 
-export type AccountTransactionPayload = SimpleTransfer;
+export interface SimpleTransferWithMemoPayload extends SimpleTransferPayload {
+    /** The provided memo of the transaction,
+       * expected to be a HEX string representing the bytes of the memo */
+    memo: string;
```
Comment (orhoj) [inline src/types.ts:471]: I think the typing should be a `Buffer` and not the visual representation of it (a hex string).

File: src/serializationHelpers.ts [hunk 3]
```diff
@@ -49,6 +49,22 @@ export function encodeWord32(value: number): Buffer {
     return Buffer.from(new Uint8Array(arr));
 }
 
+/**
+ * Encodes a 16 bit unsigned integer to a Buffer using big endian.
+ * @param value a 16 bit integer
+ * @returns big endian serialization of the input
+ */
+export function encodeWord16(value: number): Buffer {
+    if (value > 65535 || value < 0) {
```
Comment (orhoj) [inline src/serializationHelpers.ts:58]: I guess in `encodeWord16` and `encodeWord32` that we should also check `Number.isInteger()`.

File: src/serializationHelpers.ts [hunk 4]
```diff
@@ -57,3 +73,14 @@ export function encodeWord32(value: number): Buffer {
 export function encodeUint8(value: number): Buffer {
     return Buffer.from(Buffer.of(value));
 }
+
+/**
+ * Encodes a memo.
+ * @param memo a HEX string containing the bytes of the memo.
+ * @returns CBOR encoded serialization of the input.
```
Comment (orhoj) [inline src/serializationHelpers.ts:80]: This does not have anything to do with CBOR encoding.

File: src/accountTransactions.ts [hunk 5]
```diff
@@ -1,45 +1,58 @@
...
 
-    const handler = accountTransactionHandlerMap.get(type);
-    if (!handler) {
-        throw new Error(
-            'The handler map is missing the provided type: ' + type
+    serialize(transfer: SimpleTransferWithMemoPayload): Buffer {
+        const serializedToAddress = transfer.toAddress.decodedAddress;
+        const serializedAmount = encodeWord64(transfer.amount.microGtuAmount);
+        const serializedMemo = encodeMemo(
+            (transfer as SimpleTransferWithMemoPayload).memo
```
Comment (orhoj) [inline src/accountTransactions.ts:36]: `transfer as SimpleTransferWithMemoPayload` seems redundant when you have
```
serialize(transfer: SimpleTransferWithMemoPayload)
```

File: src/serializationHelpers.ts (hunk 3 above)
Comment (shjortConcordium) [reply src/serializationHelpers.ts:58]: I guess we should also add the checks for encodeWord8

File: src/accountTransactions.ts (hunk 5 above)
Comment (shjortConcordium) [reply src/accountTransactions.ts]: It is, I was also able to refactor more away.

File: README.md (hunk 1 above)
Comment (orhoj) [reply README.md]: I don't see that you made any changes to this.

File: src/accountTransactions.ts [hunk 6]
```diff
@@ -1,45 +1,51 @@
...
-        );
+) {
+    switch (type) {
+        case AccountTransactionType.SimpleTransfer:
+            return new SimpleTransferHandler();
+        case AccountTransactionType.SimpleTransferWithMemo:
+            return new SimpleTransferWithMemoHandler();
+        default:
+            throw new Error(
+                'The handler map is missing the provided type: ' + type
```
Comment (orhoj) [inline src/accountTransactions.ts:48]: This references a map that does not exist anymore. So update the error message.

File: src/accountTransactions.ts [hunk 7]
```diff
@@ -1,45 +1,63 @@
...
 }
 
+export class SimpleTransferWithMemoHandler
+    extends SimpleTransferHandler
+    implements AccountTransactionHandler<SimpleTransferWithMemoPayload>
+{
+    serialize(transfer: SimpleTransferWithMemoPayload): Buffer {
+        const regularPayload = super.serialize(transfer);
+        const serializedMemo = encodeMemo(transfer.memo);
+        return Buffer.concat([regularPayload, serializedMemo]);
```
Comment (orhoj) [inline src/accountTransactions.ts:38]: This serialization is incorrect. It has to be address, memo and then amount.

File: src/types.ts [hunk 8]
```diff
@@ -453,15 +457,22 @@ export interface AccountTransactionHeader {
...
 
-    /** the recipient of the transfer*/
+    /** the recipient of the transfer */
     toAddress: AccountAddress;
 }
 
-export type AccountTransactionPayload = SimpleTransfer;
+export interface SimpleTransferWithMemoPayload extends SimpleTransferPayload {
+    /** The bytes representation of the memo of the transaction  */
+    memo: Buffer;
```
Comment (orhoj) [inline src/types.ts:470]: To stay consistent with what we have done for other fields, we should introduce a `Memo` class that validates that the memo has a valid size. With the current implementation I won't know that my transaction is invalid until I send it to a node.

Comment (shjortConcordium) [reply src/types.ts]: Added such a class

**Pull request merged**